File: ts/src/adapters/cursor.ts
import { register } from '../registry.js'
import { HarnessError, finalizeCommand, validateRunSpec } from '../base.js'
import type { Adapter, BuildCommand, ParsedOutput, RunSpec, SubprocOutcome } from '../base.js'

// Cursor CLI `agent --print --output-format stream-json` writes one JSON
// object per stdout line (verified against 2026.09.02-c22c1a3). The stream
// carries system/user/assistant/tool_call events and a final `result` event
// with duration, result text, session_id and an optional usage block whose
// inputTokens already excludes cache reads/writes. Only that last result's
// input/output counts are reported; cache metrics are not folded in, no USD
// total is reported, and the events are preserved verbatim, in order.

type JsonObject = Record<string, unknown>

function isJsonObject(value: unknown): value is JsonObject {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

const cursorAdapter: Adapter = {
  name: 'cursor',
  instructionsFilename: 'AGENTS.md',
  defaultModel: '', // Leave model selection to the caller's Cursor config.
  permissionBypassArgs: ['--force'],
  configHomeEnv: 'CURSOR_CONFIG_DIR',
  // Native headless mode registers a SIGINT AbortController that aborts all
  // work; SIGTERM only triggers the global exit path.
  gracefulSignal: 'SIGINT',

  buildCommand(spec: RunSpec): BuildCommand {
    const validated = validateRunSpec(this, spec)
    if (!spec.prompt) {
      throw new HarnessError('cursor requires a non-empty prompt for non-interactive mode', 'invalid-options')
    }
    const { model, permissionArgs } = validated
    const args = ['--print', '--output-format', 'stream-json', '--stream-partial-output']
    if (model) args.push('--model', model)
    // `--` keeps leading '-' prompts from being parsed as flags.
    args.push(...permissionArgs, '--', spec.prompt)
    return finalizeCommand(this, spec, validated, { cmd: 'agent', args, model: spec.model || null })
  },

  parseOutput(_spec: RunSpec, outcome: SubprocOutcome): ParsedOutput {
    const events: JsonObject[] = []
    // A complete final line needs no trailing newline; a truncated one fails to
    // parse and is dropped like any other malformed line. JSON.parse already
    // rejects the non-standard NaN/Infinity constants.
    for (const line of outcome.stdout.split('\n')) {
      const s = line.trim()
      if (!s) continue
      let parsed: unknown
      try {
        parsed = JSON.parse(s)
      } catch {
        continue
      }
      if (isJsonObject(parsed)) events.push(parsed)
    }
    let tokensIn: number | null = null
    let tokensOut: number | null = null
    for (let i = events.length - 1; i >= 0; i--) {
      const event = events[i]!
      if (event['type'] !== 'result') continue
      const usage = event['usage']
      if (isJsonObject(usage)) {
        // Counts must be exact in both language APIs; reject unsafe integers
        // rather than silently report a rounded token total.
        const input = usage['inputTokens']
        const output = usage['outputTokens']
        tokensIn = typeof input === 'number' && Number.isSafeInteger(input) && input >= 0 ? input : null
        tokensOut = typeof output === 'number' && Number.isSafeInteger(output) && output >= 0 ? output : null
      }
      break
    }
    return { costUsd: null, tokensIn, tokensOut, raw: events.length === 0 ? null : events }
  },
}

cursorAdapter.installMeta = {
  packageManager: 'binary',
  installCommand: ['bash', '-c', 'curl https://cursor.com/install -fsS | bash'],
  updateCommand: ['agent', 'update'],
  versionCommand: ['agent', '--version'],
  platforms: ['darwin', 'linux'],
}

register('cursor', cursorAdapter)

File: src/harness/adapters/cursor.py
"""Cursor CLI (`agent`) in headless stream-JSON mode.

`agent --print --output-format stream-json --stream-partial-output -- <prompt>`
emits one JSON object per stdout line (JSONL). Qualified with 2026.09.02-c22c1a3,
the final `result` event carries `duration_ms`, `result`, `session_id` and an
optional `usage` object whose `inputTokens` (already net of cache reads and
writes upstream) and `outputTokens` supply the token metrics; cache counters
and partial assistant deltas are never summed. No dollar total is reported, so
cost stays None. Every valid JSON object event is preserved in `raw` (None when
the stream carried none).

The official installer (`curl https://cursor.com/install | bash`) provides the
`agent` binary; the legacy `cursor-agent` name accepts the same arguments and
works through `RunSpec.executable`. Tool permissions retain upstream policy;
`--force` is the explicit bypass. Harness adds no separate trust, MCP approval
or sandbox flags. Sessions, resume and worker modes are unsupported. Headless
mode aborts in-flight work on SIGINT but treats SIGTERM as a global exit, so
SIGINT is the graceful signal.
"""
from __future__ import annotations

import json
import math

from harness._subproc import SubprocOutcome
from harness.base import Adapter, BuildCommand, HarnessError, InstallMeta, ParsedOutput, RunSpec


class CursorAdapter(Adapter):
    name = "cursor"
    instructions_filename = "AGENTS.md"
    DEFAULT_MODEL = ""  # Leave model selection to the caller's Cursor config.
    permission_bypass_args = ("--force",)
    config_home_env = "CURSOR_CONFIG_DIR"
    graceful_signal = "SIGINT"
    install_meta = InstallMeta(
        package_manager="binary",
        install_command=("bash", "-c", "curl https://cursor.com/install -fsS | bash"),
        update_command=("agent", "update"),
        version_command=("agent", "--version"),
        platforms=("darwin", "linux"),
    )

    def reported_model(self, spec: RunSpec) -> str | None:
        return spec.model or None

    def build_command(self, spec: RunSpec) -> BuildCommand:
        resolved = self.resolve_run_spec(spec)
        if not spec.prompt:
            raise HarnessError("cursor requires a non-empty prompt for headless mode", code="invalid-options")
        args = ["--print", "--output-format", "stream-json", "--stream-partial-output"]
        if resolved.model:
            args.extend(["--model", resolved.model])
        args.extend(resolved.permission_args)
        # '--' keeps leading '-' prompts from being parsed as flags.
        args.extend(["--", spec.prompt])
        return self.finalize_command(spec, cmd="agent", args=args)

    def parse_output(self, spec: RunSpec, outcome: SubprocOutcome) -> ParsedOutput:
        events = _json_object_events(outcome.stdout)
        tokens_in, tokens_out = _result_usage(events)
        return {"cost_usd": None, "tokens_in": tokens_in, "tokens_out": tokens_out, "raw": events}


def _reject_constant(value: str) -> None:
    raise ValueError(f"non-JSON constant: {value}")


def _json_object_events(stdout: str) -> list[dict] | None:
    """Every complete JSON object line, in order; malformed, truncated and
    non-object lines are skipped. A final line without a newline still counts."""
    events: list[dict] = []
    for line in stdout.split("\n"):
        line = line.strip()
        if not line:
            continue
        try:
            event = json.loads(line, parse_constant=_reject_constant)
        except ValueError:
            continue
        if isinstance(event, dict):
            events.append(event)
    return events or None


def _result_usage(events: list[dict] | None) -> tuple[int | None, int | None]:
    """`inputTokens` / `outputTokens` from the last `result` event's `usage`,
    each validated independently; missing or invalid counters are None."""
    if not events:
        return None, None
    for event in reversed(events):
        if event.get("type") != "result":
            continue
        usage = event.get("usage")
        if not isinstance(usage, dict):
            return None, None
        return _token_count(usage.get("inputTokens")), _token_count(usage.get("outputTokens"))
    return None, None


def _token_count(value: object) -> int | None:
    """A nonnegative integer exactly representable in both language APIs."""
    if isinstance(value, bool):
        return None
    if isinstance(value, int):
        return value if 0 <= value <= 2**53 - 1 else None
    if isinstance(value, float) and 0 <= value <= 2**53 - 1 and math.isfinite(value) and value.is_integer():
        return int(value)
    return None
